primitives/src/pcs/univariate_kzg/mod.rs

```diff
@@ -652,6 +651,344 @@ where
+            let bases: Vec<IcicleAffine<C>> =
+                parallelizable_slice_iter(&prover_param.powers_of_g[..supported_degree + 1])
+                    .map(|&p| Self::ark_affine_to_icicle(p))
+                    .collect();
+            #[cfg(feature = "kzg-print-trace")]
+            end_timer!(conv_time);
+
+            #[cfg(feature = "kzg-print-trace")]
+            let load_time = start_timer!(|| "Load group elements: CPU->GPU");
+            bases_on_device.copy_from_host(&bases)?;
+            #[cfg(feature = "kzg-print-trace")]
+            end_timer!(load_time);
+
+            Ok(bases_on_device)
+        }
+
+        /// Load polynomial's coefficients onto GPU, preparing for poly-commit
+        /// on GPU later
+        fn load_poly_to_gpu<'poly>(
+            poly: &DensePolynomial<E::ScalarField>,
+        ) -> Result<HostOrDeviceSlice<'poly, C::ScalarField>, PCSError> {
+            let size = poly.degree() + 1;
+            let mut scalars_on_device = HostOrDeviceSlice::<'_, C::ScalarField>::cuda_malloc(size)?;
+
+            #[cfg(feature = "kzg-print-trace")]
+            let conv_time = start_timer!(|| "Type Conversion: ark->ICICLE: Scalar");
+            let scalars: Vec<C::ScalarField> = parallelizable_slice_iter(&poly.coeffs()[..size])
+                .map(|&s| Self::ark_field_to_icicle(s))
+                .collect();
+            #[cfg(feature = "kzg-print-trace")]
+            end_timer!(conv_time);
+
+            #[cfg(feature = "kzg-print-trace")]
+            let load_time = start_timer!(|| "Load scalars: CPU->GPU");
+            scalars_on_device.copy_from_host(&scalars)?;
+            #[cfg(feature = "kzg-print-trace")]
+            end_timer!(load_time);
+
+            Ok(scalars_on_device)
+        }
+
+        /// Similar to [`Self::load_poly_to_gpu()`] but handling a batch of poly
+        /// of the same degree at once
+        fn load_batch_poly_to_gpu<'poly>(
+            polys: &[DensePolynomial<E::ScalarField>],
+        ) -> Result<HostOrDeviceSlice<'poly, C::ScalarField>, PCSError> {
+            if polys.is_empty() {
+                return Err(PCSError::InvalidParameters(
+                    "number of polys must be positive".to_string(),
+                ));
+            }
+
+            let size = polys[0].degree() + 1;
+            if polys.iter().any(|p| p.degree() + 1 != size) {
+                return Err(PCSError::InvalidParameters(
+                    "all polys should have the same degree".to_string(),
+                ));
+            }
+
+            let mut scalars_on_device =
+                HostOrDeviceSlice::<'_, C::ScalarField>::cuda_malloc(size * polys.len())?;
+
+            #[cfg(feature = "kzg-print-trace")]
+            let conv_time = start_timer!(|| "Type Conversion: ark->ICICLE: Scalar");
+            let scalars: Vec<C::ScalarField> = parallelizable_slice_iter(polys)
+                .map(|poly| {
```

Comment:
nit: we could directly use `flat_map` here.

Also according to profiling, this is taking a long time committing 10 degree 2^20 polys.
We should have an issue for faster type conversion

Reply:
I suspect that this step takes some unnecessary mem copies

Reply:
Don't know why but the following code has 2x speed up
```rust
           #[cfg(feature = "kzg-print-trace")]
            let conv_time = start_timer!(|| "Type Conversion 1: ark->ICICLE: Scalar");
            let escalars: Vec<E::ScalarField> = polys
                .iter()
                .flat_map(|poly| poly.coeffs())
                .copied()
                .collect();
            let _scalars: Vec<C::ScalarField> = parallelizable_slice_iter(&escalars)
                .map(|&s| Self::ark_field_to_icicle(s))
                .collect();
            #[cfg(feature = "kzg-print-trace")]
            end_timer!(conv_time);
```

Reply:
confirm that this is a problem, scalar conversion time scales more than linearly with the batch_size. there's definitely some unnecessary memory copying.

Reply:
implemented your `flat_map` idea in [7e5862c](https://github.com/EspressoSystems/jellyfish/pull/498/commits/7e5862c2d161bacca594216404465f94cae04a8f)! 
conversion time for scalars drop from 250ms to 108ms
(and MSM took 282ms, so roughly 1/3 of that, and roughly 1/4 of overall runtime)

the single instance takes 5.8 ms. so we are still suffer a super-linear cost (but much better now!)

Reply:
more experiment: a single MSM of 2^23, and batch MSM 8 * 2^20
turns out the scalar conversion of latter is exactly 2x of the former, all the rest is the same. 

this should give the definitive reason: the latter have exactly 1 extra memory copy which cause the 2x runtime. and this copy happens during the two `.collect()`. 

unfortunately, rayon's `ParallelIterator` can only be created from collections or slices, not directly from `Iterator` from std. 

IMO, this experiment demonstrates that we have squeezed the performance as much as we could without violating Rust's type checking. and we should accept this and close #507 now. 
@mrain

Reply:
Check this: https://docs.rs/rayon/latest/rayon/iter/trait.ParallelBridge.html

There is still room for improvement. I can investigate it later.